From 6e390a19b3389e17e430635738085c8ac2f1ed94 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Fri, 28 Jan 2022 05:02:38 +0200
Subject: [PATCH] [SW-68803] gaudi3: port security functions

Porting image verification and dependent functions (key store and
load) for gaudi3. This is in preparation for gaudi3 to use image
descriptor for image verification.

Port image verification and dependent functions for gaudi3.

secureboot_verify_image_v2 : New api for signature verification.
secureboot_verify_image : Dummy function maintained for common code
compilation.

Change-Id: I7eed728f91cddff2d6737418fea5bd00ca551d80
---

diff --git a/arch/gaudi2/arc/secureboot.c b/arch/gaudi2/arc/secureboot.c
index 009c1cb..851c5ce 100644
--- a/arch/gaudi2/arc/secureboot.c
+++ b/arch/gaudi2/arc/secureboot.c
@@ -70,7 +70,7 @@
 	}
 };
 
-static int security_check_key_revoke_status(uint32_t key_index)
+int security_check_key_revoke_status(uint32_t key_index)
 {
 	uint8_t max_key_index = NUM_KEYS_IN_ROM, key_revoke;
 	enum security_key_location key_location;
diff --git a/arch/gaudi3/arc/dev_security.h b/arch/gaudi3/arc/dev_security.h
index 719c06f..916b130 100644
--- a/arch/gaudi3/arc/dev_security.h
+++ b/arch/gaudi3/arc/dev_security.h
@@ -8,13 +8,11 @@
 
 #include <sha512.h>
 
-/* TODO: Replace with gaudi3 values */
-
 #define NUM_KEYS_IN_ROM					5
 #define NUM_KEYS_IN_EFUSE				3
 #define SECURITY_DWORD_PER_KEY				12
 #define SECURITY_KEY_HASH_COUNT				3
-#define NUM_DWORDS_AVLBL_IN_PSOC_SEC			8
+#define SECURITY_KEY_HASH_LEN			SHA384_SUM_LEN
 
 enum security_key_location {
 	KEY_LOCATION_ROM,
@@ -39,4 +37,8 @@
 
 extern const uint8_t rom_key_hashes[NUM_KEYS_IN_ROM][SHA384_SUM_LEN];
 
+int secureboot_verify_image_v2(const uint8_t *image, uint32_t image_len,
+		uint8_t *key, uint32_t key_len,
+		uint8_t *sig, uint32_t sig_len, uint8_t sec_ver);
+
 #endif /* _GAUDI3_DEV_SECURITY_H_ */
diff --git a/arch/gaudi3/arc/secureboot.c b/arch/gaudi3/arc/secureboot.c
index b030246..1a05da5 100644
--- a/arch/gaudi3/arc/secureboot.c
+++ b/arch/gaudi3/arc/secureboot.c
@@ -5,8 +5,87 @@
  */
 
 #include <arcboot.h>
+#include <string.h>
+#include <efuse.h>
+#include <secureboot.h>
+#include <security.h>
+#include <dev_security.h>
+#include <rsa.h>
+#include <sha512.h>
 
-int secureboot_verify_image(__unused uint32_t image_size, __unused uint32_t image_offset)
+int secureboot_verify_key_hash(int key_index, uint8_t *key, uint32_t key_len)
+{
+	uint8_t expected_hash[SHA384_SUM_LEN];
+	uint8_t actual_hash[SHA384_SUM_LEN];
+
+	if (security_get_key_hash((uint32_t *)expected_hash, key_index))
+		return SECUREBOOT_VERIFY_KEY_FAILED;
+
+	sha384_csum(key, key_len, actual_hash, 0);
+
+	return memcmp(actual_hash, expected_hash, SHA384_SUM_LEN);
+}
+
+/*
+ * arcboot - Signature Verification
+ *
+ * Both the protected data and the signature of the protected data can be
+ * changed, but the public key should be locked into the device.
+ *
+ * @arguments:
+ * image     - pointer to blob of protected data
+ * image_len - size of the blob (Image + public key (inc key header))
+ * sig       - pointer to signature of the blob
+ * sig_len   - signature size
+ * key       - pointer to public key
+ * key_len   - public key size + key header (784 + 8 = 792 bytes)
+ *
+ * @return:
+ * SECUREBOOT_VERIFY_SUCCESS    : Signature validation successful
+ * SECUREBOOT_VERIFY_KEY_REVOKED: Selected key index is revoked
+ * SECUREBOOT_VERIFY_KEY_FAILED : Key hash verification failed
+ * SECUREBOOT_VERIFY_FAILED     : Signature validation failed
+ */
+int secureboot_verify_image_v2(const uint8_t *image, uint32_t image_len, uint8_t *key,
+				uint32_t key_len, uint8_t *sig, uint32_t sig_len,
+				uint8_t sec_ver)
+{
+	int ret, key_index;
+
+	if (!security_efuse_image_validation_en())
+		return SECUREBOOT_VERIFY_SUCCESS;
+
+	if ((key_len > TOTAL_PUB_KEY_PROPS_SZ) || (sig_len > SIG_SIZE))
+		return SECUREBOOT_VERIFY_FAILED;
+
+	if (security_check_valid_fw_version(sec_ver)
+			!= SECURITY_FW_VERSION_SECURE) {
+		return SECUREBOOT_VERIFY_VER_FAILED;
+	}
+
+	key_index = *(uint32_t*) (key + key_len - KEY_PROP_HDR_SZ);
+
+	/* TODO: Confirm if key revocation failure needs a dedicated status bit */
+	ret = security_check_key_revoke_status(key_index);
+	if (ret != 1)
+		return SECUREBOOT_VERIFY_KEY_REVOKED;
+
+	if (secureboot_verify_key_hash(key_index, key, key_len))
+		return SECUREBOOT_VERIFY_KEY_FAILED;
+
+	ret = rsa_verify(image, image_len, key, key_len, sig, sig_len);
+
+	/* 0 - SUCCESS, but arcboot main code, expects 1 for SUCCESS */
+	if (!ret)
+		return SECUREBOOT_VERIFY_SUCCESS;
+
+	return SECUREBOOT_VERIFY_FAILED;
+}
+
+/*
+ * Dummy function.
+ */
+int secureboot_verify_image(__unused uint32_t image_size,__unused uint32_t image_offset)
 {
 	return 0;
 }
diff --git a/arch/gaudi3/arc/security.c b/arch/gaudi3/arc/security.c
index c1f447e..3d6c0f8 100644
--- a/arch/gaudi3/arc/security.c
+++ b/arch/gaudi3/arc/security.c
@@ -5,15 +5,303 @@
  */
 
 #include <arcboot.h>
+#include "security.h"
+#include "efuse.h"
+#include "security.h"
+#include "spi_flash.h"
+#include <string.h>
 
 /* TODO: Add gaudi3 specific implementation */
+
+/* TODO: Update ROM keys for gaudi3*/
+/*
+ * This key hash is calculated on the Public Key Properties -
+ *   -----------------------------------------------
+ * | R ^ 2 | modulus | exponent | inverse | num bits |
+ *   -----------------------------------------------
+ * Since the key props are directly given to the ROM, there is
+ * no need to explicitly include the Public key, which is
+ * usually the modulus.
+ */
+const uint8_t rom_key_hashes[NUM_KEYS_IN_ROM][SHA384_SUM_LEN] = {
+	/* Habana Public Key */
+	{
+		0x08, 0x51, 0xe7, 0x07, 0xc1, 0x0a, 0xc8, 0x45,
+		0x51, 0x61, 0xb4, 0xdf, 0xb6, 0xd6, 0x89, 0xbc,
+		0x57, 0xbb, 0xf9, 0xe2, 0x50, 0x4f, 0x44, 0x1f,
+		0x12, 0x4a, 0x73, 0x3d, 0xbb, 0xda, 0x63, 0xf2,
+		0x76, 0x53, 0x6c, 0x12, 0xa1, 0x13, 0x37, 0x30,
+		0x03, 0xee, 0x79, 0xce, 0x20, 0xa3, 0x32, 0x33
+	},
+	/* Additional Keys*/
+	/* Production Key 1 */
+	{
+		0x47, 0x9b, 0x51, 0x5d, 0xf9, 0x8b, 0xfd, 0x3d,
+		0x38, 0x3f, 0x19, 0x51, 0x40, 0x62, 0xcb, 0x15,
+		0x2d, 0xae, 0x11, 0xe8, 0x1a, 0x91, 0x3e, 0x02,
+		0xa0, 0x80, 0x08, 0x9a, 0x62, 0x93, 0xb7, 0x65,
+		0x73, 0xe1, 0x6f, 0x86, 0x32, 0x16, 0xea, 0x37,
+		0x48, 0x19, 0xb8, 0xe0, 0xc3, 0x3b, 0x5a, 0xcf
+	},
+	/* Production Key 2 */
+	{
+		0x48, 0x3c, 0x50, 0xfc, 0xf1, 0x3a, 0x56, 0x7a,
+		0x7b, 0x7c, 0xcc, 0x77, 0x66, 0xb3, 0x85, 0x1b,
+		0x6f, 0x43, 0x60, 0xce, 0x16, 0x9a, 0xc4, 0x14,
+		0xd5, 0xcb, 0x11, 0x09, 0x66, 0xc0, 0xfe, 0x42,
+		0xfe, 0x78, 0x79, 0x57, 0xdb, 0x06, 0x5c, 0x37,
+		0x2f, 0xe2, 0x81, 0xb0, 0x00, 0x11, 0x9f, 0x8e,
+	},
+	/* Production Key 3 */
+	{
+		0x5c, 0xc8, 0xd6, 0xf3, 0x42, 0xdc, 0xa3, 0xdf,
+		0xb0, 0xa3, 0xa8, 0x73, 0x68, 0x54, 0x78, 0x5e,
+		0x8c, 0xaf, 0x09, 0x0b, 0x0d, 0x41, 0xa8, 0xa1,
+		0xd9, 0xf8, 0x31, 0xdc, 0x95, 0x0f, 0x84, 0x02,
+		0xb3, 0x2a, 0x61, 0xe4, 0x10, 0xd8, 0x39, 0xdf,
+		0x80, 0x27, 0xc1, 0x2e, 0xd5, 0xd2, 0xb2, 0xcd,
+	},
+	/* Production Key 4 */
+	{
+		0xdb, 0x8c, 0x8c, 0xe7, 0x08, 0x19, 0x6b, 0x04,
+		0x76, 0x43, 0xa7, 0xe4, 0x34, 0x2d, 0x2e, 0x8f,
+		0xf5, 0xdd, 0x20, 0xc9, 0xc7, 0x48, 0x0c, 0x4d,
+		0x80, 0x35, 0x18, 0x10, 0x8b, 0xb2, 0x04, 0xd2,
+		0x79, 0xc1, 0x6b, 0x33, 0x07, 0x99, 0x21, 0x3a,
+		0xbf, 0x77, 0xfa, 0x6b, 0x53, 0xf4, 0x27, 0x25,
+	}
+};
+
+int security_get_key_hash(uint32_t *hash, uint8_t key_index)
+{
+	uint32_t valid_key_base, max_key_index;
+	int i;
+	enum security_key_location key_location;
+	uint32_t *ptr;
+
+	key_location = ((arcboot_rom_global.config_fuse >>
+			EFUSE_SYSCFG_KEY_LOCATION) &
+			EFUSE_SYSCFG_KEY_LOCATION_MASK);
+
+	/* check efuse for key location */
+	if (key_location == KEY_LOCATION_EFUSE) {
+		max_key_index = NUM_KEYS_IN_EFUSE;
+
+		if (key_index >= max_key_index)
+			return 1;
+
+		/* read from efuse/SP SRAM */
+		valid_key_base = (key_index * SECURITY_KEY_HASH_LEN) + SECURITY_EFUSE_KEYS_OFFSET;
+
+		for (i = 0; i < SECURITY_DWORD_PER_KEY; i++)
+			*(hash + i) = sp_sram_readl(valid_key_base + (i *
+					SECURITY_BYTE_PER_DWORD));
+	} else {
+		/*
+		 * All other cases are routed into ROM, since it is
+		 * the more secure way
+		 */
+		max_key_index = NUM_KEYS_IN_ROM;
+
+		if (key_index >= max_key_index)
+			return 1;
+
+		/* read from ROM itself */
+		ptr = (uint32_t *)rom_key_hashes[key_index];
+		memcpy(hash, ptr, SECURITY_KEY_HASH_LEN);
+	}
+
+	return 0;
+}
+
+int security_check_key_revoke_status(uint32_t key_index)
+{
+	uint8_t max_key_index = NUM_KEYS_IN_ROM, key_revoke;
+	enum security_key_location key_location;
+
+	key_location = ((arcboot_rom_global.config_fuse >>
+			EFUSE_SYSCFG_KEY_LOCATION) &
+			EFUSE_SYSCFG_KEY_LOCATION_MASK);
+	/*
+	 * key location = EFUSE (1)
+	 * key location = ROM (0, 2, 3)
+	 * Key index ranges from:
+	 * 0 - 4(5 keys): if key location == ROM
+	 * 0 - 2(3 keys): if key location == eFUSE
+	 * Except a value of 1 for EFUSE, rest all are
+	 * routed into ROM, since thats the secure way.
+	 */
+	/* check efuse for key location */
+	if (key_location == KEY_LOCATION_EFUSE)
+		max_key_index = NUM_KEYS_IN_EFUSE;
+
+	if (key_index >= max_key_index)
+		return -1;
+
+	key_revoke = sp_sram_readl(SECURITY_EFUSE_KEYS_REVOKE_OFFSET);
+
+	/* if the key is not revoked, return 1 */
+	if ((key_revoke & BIT(key_index)) == 0)
+		return 1;
+
+	/* no valid key found */
+	return -1;
+}
+
+static void security_rom_get_hash(uint32_t *key_hash)
+{
+	uint32_t *ptr = (uint32_t *)rom_key_hashes;
+	int size = NUM_KEYS_IN_ROM * SECURITY_DWORD_PER_KEY;
+
+	for (int i = 0; i < size; i++)
+		*(key_hash + i) = *(ptr + i);
+}
 
 void security_i2c_s_rr_hbw_config(__unused uint8_t spi_controller_en) {}
 void security_smi_rr_hbw_config(__unused uint8_t spi_controller_en) {}
 void security_jtag_rr_hbw_config(__unused uint8_t spi_controller_en) {}
 void security_disable_jtag_if(void) {}
 void security_handle_twi(void) {}
-void security_load_keys(void) {}
+
+/* TODO: get newly added key addresses from global specs once it is updated */
+#define mmD0_PSOC_SECURITY_PUBLIC_KEY_3_0		0xC494190
+#define mmD0_PSOC_SECURITY_PUBLIC_KEY_4_0		0xC4941C0
+static void security_store_keys_in_psoc(uint8_t max_keys, uint32_t *hash)
+{
+	uint8_t key_idx, dword;
+	uint32_t key_soc_location_base;
+	uint32_t key_address[] = {
+		mmD0_PSOC_SECURITY_PUBLIC_KEY_0_0,
+		mmD0_PSOC_SECURITY_PUBLIC_KEY_1_0,
+		mmD0_PSOC_SECURITY_PUBLIC_KEY_2_0,
+		mmD0_PSOC_SECURITY_PUBLIC_KEY_3_0,
+		mmD0_PSOC_SECURITY_PUBLIC_KEY_4_0
+	};
+
+	for (key_idx = 0; key_idx < max_keys; key_idx++) {
+		key_soc_location_base = key_address[key_idx];
+		for (dword = 0; dword < SECURITY_DWORD_PER_KEY; dword++) {
+			regmap_writel(hash[(key_idx * SECURITY_DWORD_PER_KEY) + dword],
+					key_soc_location_base);
+			key_soc_location_base += sizeof(uint32_t);
+		}
+	}
+}
+
+void security_load_keys(void)
+{
+	uint8_t hash_key_index, num_dwords, max_keys = NUM_KEYS_IN_ROM;
+	uint32_t key_hash[NUM_KEYS_IN_ROM * SECURITY_DWORD_PER_KEY];
+	enum security_key_location key_location;
+
+	key_location = ((arcboot_rom_global.config_fuse >>
+			EFUSE_SYSCFG_KEY_LOCATION) &
+			EFUSE_SYSCFG_KEY_LOCATION_MASK);
+
+	arcboot_rom_status_update(PROGRESS_KEY_LOC);
+	arcboot_rom_status_update_location(PROGRESS_KEY_LOC_0, key_location);
+
+	switch (key_location) {
+	case KEY_LOCATION_EFUSE:
+		max_keys = NUM_KEYS_IN_EFUSE;
+		efuse_get_hash(key_hash);
+		break;
+	case KEY_LOCATION_ROM:
+	default:
+		security_rom_get_hash(key_hash);
+		break;
+	}
+
+	/* store hashes in the limited space available in PSOC_SECURITY */
+	security_store_keys_in_psoc(max_keys, key_hash);
+
+	/* Store key hashes in SPSRAM */
+	num_dwords = max_keys * SECURITY_DWORD_PER_KEY;
+	for (hash_key_index = 0; hash_key_index < num_dwords;
+					hash_key_index++)
+		sp_sram_writel(key_hash[hash_key_index],
+			   SECURITY_EFUSE_KEYS_OFFSET + (hash_key_index *
+			   SECURITY_BYTE_PER_DWORD));
+}
+
+void security_load_revoke_bits(void)
+{
+	uint32_t key_revoke = 0;
+	enum security_location revoke_location;
+
+	revoke_location = ((arcboot_rom_global.config_fuse >>
+			   EFUSE_SYSCFG_REVOKATION_LOCATION) &
+			   EFUSE_SYSCFG_REVOKATION_LOCATION_MASK);
+
+	arcboot_rom_status_update(PROGRESS_REVOKE_LOC);
+	arcboot_rom_status_update_location(PROGRESS_REVOKE_LOC_0,
+					   revoke_location);
+
+	switch (revoke_location) {
+	/* TODO: Add other locations if required for gaudi3 */
+	/* TODO: Confirm enumeration values for gaudi3 */
+	case ASIC_EFUSE:
+		key_revoke = arcboot_rom_global.config_fuse;
+		break;
+	default:
+		return;
+	}
+	key_revoke &= EFUSE_SYSCFG_REVOKE_KEY_MASK;
+
+	/* store key_revoke map in PSOC */
+	regmap_writel(key_revoke, mmSECURITY_KEY_REVOKE);
+
+	/* store key hashes in Scratch PAD Sram */
+	sp_sram_writel(key_revoke,
+		       SECURITY_EFUSE_KEYS_REVOKE_OFFSET);
+}
+
+void security_load_fw_version(void)
+{
+	uint8_t fw_version_sec = 0;
+	enum security_location fw_version_location;
+
+	fw_version_location = ((arcboot_rom_global.config_fuse >>
+			       EFUSE_SYSCFG_FW_VERSION_LOCATION) &
+			       EFUSE_SYSCFG_FW_VERSION_LOCATION_MASK);
+
+	arcboot_rom_status_update(PROGRESS_VER_LOC);
+	arcboot_rom_status_update_location(PROGRESS_VER_LOC_0,
+					   fw_version_location);
+
+	switch (fw_version_location) {
+	/* TODO: Add other locations if required for gaudi3 */
+	/* TODO: Confirm enumeration values for gaudi3 */
+	case ASIC_EFUSE:
+		fw_version_sec = efuse_otp_get_fw_version_security();
+		break;
+	default:
+		return;
+	}
+
+	/* store minimum allowed security version in PSOC */
+	regmap_writel(SECURITY_FW_VERSION_FROM_SECURITY_BYTE(fw_version_sec),
+		      mmSECURITY_MIN_FW_VERSION);
+
+	/* store minimum allowed security version in Scratch PAD Sram */
+	sp_sram_writel(SECURITY_FW_VERSION_FROM_SECURITY_BYTE(fw_version_sec),
+		      SECURITY_EFUSE_FW_VERSION_OFFSET);
+}
+
+void security_load_sec_efuse_dword(void)
+{
+	sp_sram_writel(arcboot_rom_global.config_fuse,
+		       SECURITY_EFUSE_SYS_CFG_OFFSET);
+}
+
+int security_efuse_image_validation_en(void)
+{
+	/* check efuse for security validation enable */
+	uint32_t config_fuse = sp_sram_readl(SECURITY_EFUSE_SYS_CFG_OFFSET);
+
+	return !!(config_fuse & EFUSE_SYSCFG_ENABLE_VALIDATION_MASK);
+}
+
 void security_enable_jtag_if(void) {}
 void security_jtag_dap(__unused uint8_t status) {}
 void check_and_enable_4b_termination(void) {}
diff --git a/common/secureboot.h b/common/secureboot.h
index 313e6b0..b655579 100644
--- a/common/secureboot.h
+++ b/common/secureboot.h
@@ -15,6 +15,7 @@
 
 #define SECUREBOOT_HASH_BLOCK_CALC_BUF	512
 
+#define SECUREBOOT_VERIFY_KEY_REVOKED	-3
 #define SECUREBOOT_VERIFY_KEY_FAILED	-2
 #define SECUREBOOT_VERIFY_VER_FAILED	-1
 #define SECUREBOOT_VERIFY_FAILED	0
diff --git a/common/security.c b/common/security.c
index 1ccfcaa..3048006 100644
--- a/common/security.c
+++ b/common/security.c
@@ -151,7 +151,7 @@
 	return SECURITY_FW_VERSION_NOT_SECURE;
 }
 
-void security_load_revoke_bits(void)
+__weak void security_load_revoke_bits(void)
 {
 	uint32_t key_revoke = 0;
 	enum security_location revoke_location;
@@ -188,7 +188,7 @@
 		       SECURITY_EFUSE_KEYS_REVOKE_OFFSET);
 }
 
-void security_load_fw_version(void)
+__weak void security_load_fw_version(void)
 {
 	uint8_t fw_version_sec = 0;
 	enum security_location fw_version_location;
diff --git a/common/security.h b/common/security.h
index 3f6553e..d375c52 100644
--- a/common/security.h
+++ b/common/security.h
@@ -137,5 +137,7 @@
 void check_and_enable_4b_termination(void);
 void security_enable_twi_if(void);
 void security_handle_twi(void);
+int security_check_key_revoke_status(uint32_t key_index);
+int security_efuse_image_validation_en(void);
 
 #endif /* _SECURITY_H_ */
